### PR TITLE
enh(UI): Improve annotations over graphs

### PR DESCRIPTION
**Pull request by Thebarda**
## Description

This improves annotations (comment, acknowlegment and downtime) by highlighting the hovered annotation.

I handle a margin error (1px left, 1px right) when hovering a comment or an acknowledgement

https://www.loom.com/share/a15785a4544c434b98c06a03b1f54d46

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 19.10.x
- [ ] 20.04.x
- [ ] 20.10.x
- [x] 21.04.x (master)

<h2> How this pull request can be tested ? </h2>

- On Resourse Status, click on a row with graph data
- Add some comments or acknowledgements or downtimes
- Hover a annotation, it should be highlighted and the other annotations should be a bit faded off.

## Checklist

- [x] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

**Comment by bdauria** [comment]
Tested and works like a charm 👍  

I feel like we could improve the reusability of the annotationHovered conditions. Maybe by using hooks?

**Review by bdauria — COMMENTED** [review]
Tested and works like a charm 👍

I feel like we could improve the reusability of the annotationHovered conditions. Maybe by using hooks?

**Comment by Thebarda** [comment]
> I feel like we could improve the reusability of the annotationHovered conditions. Maybe by using hooks?

Should I rename the `annotationHovered` btw ?

File: www/front_src/src/Resources/Graph/Performance/Graph/useAnnotations.tsx [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    setAnnotationHovered(
+      find(({ startDate, endDate, date }: TimelineEvent) => {
+        if (isNil(startDate)) {
+          return isBetweenErorMargin(xScale(new Date(date)));
+        }
+
+        const isBetweenStartAndEndDate = getIsBetween({
+          xStart: xScale(new Date(startDate as string)),
+          xEnd: xScale(
+            endDate ? new Date(endDate) : (last(xScale.domain()) as Date),
```
Comment (bdauria) [inline www/front_src/src/Resources/Graph/Performance/Graph/useAnnotations.tsx:80]:
```suggestion
            endDate ? new Date(endDate) : last<Date>(xScale.domain()),
```

File: www/front_src/src/Resources/Graph/Performance/Graph/useAnnotations.tsx [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    React.SetStateAction<TimelineEvent | undefined>
+  >;
+  getStrokeWidth: (event: TimelineEvent) => number;
+  getStrokeOpacity: (event: TimelineEvent) => number;
+  getFill: (props: GetColor) => string;
+  getIconColor: (props: GetColor) => string;
+  changeAnnotationHovered: (props: ChangeAnnotationHovered) => void;
+}
+
+interface GetColor {
```
Comment (bdauria) [inline www/front_src/src/Resources/Graph/Performance/Graph/useAnnotations.tsx:38]:
```suggestion
interface GetColorProps {
```

File: www/front_src/src/Resources/Graph/Performance/Graph/useAnnotations.tsx [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+import { TimelineEvent } from '../../../Details/tabs/Timeline/models';
+
+export interface Annotations {
+  annotationHovered: TimelineEvent | undefined;
+  setAnnotationHovered: React.Dispatch<
+    React.SetStateAction<TimelineEvent | undefined>
+  >;
+  getStrokeWidth: (event: TimelineEvent) => number;
+  getStrokeOpacity: (event: TimelineEvent) => number;
+  getFill: (props: GetColor) => string;
```
Comment (bdauria) [inline www/front_src/src/Resources/Graph/Performance/Graph/useAnnotations.tsx:33]:
```suggestion
  getFill: (props: GetColorProps) => string;
```

File: www/front_src/src/Resources/Graph/Performance/Graph/useAnnotations.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  getIconColor: (props: GetColor) => string;
```
Comment (bdauria) [inline www/front_src/src/Resources/Graph/Performance/Graph/useAnnotations.tsx:34]:
```suggestion
  getIconColor: (props: GetColorProps) => string;
```

File: www/front_src/src/Resources/Graph/Performance/Graph/useAnnotations.tsx [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    ])(annotationHovered);
+
+  const getFill = ({ color, event }: GetColor): string =>
+    cond<TimelineEvent | undefined, string>([
+      [isNil, always(fade(color, 0.3))],
+      [equals<TimelineEvent | undefined>(event), always(fade(color, 0.5))],
+      [T, always(fade(color, 0.1))],
+    ])(annotationHovered);
+
+  const getIconColor = ({ color, event }: GetColor): string =>
```
Comment (bdauria) [inline www/front_src/src/Resources/Graph/Performance/Graph/useAnnotations.tsx:110]:
```suggestion
  const getIconColor = ({ color, event }: GetColorProps): string =>
```

File: www/front_src/src/Resources/Graph/Performance/Graph/useAnnotations.tsx [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    ])(annotationHovered);
+
+  const getStrokeOpacity = (event: TimelineEvent): number =>
+    cond<TimelineEvent | undefined, number>([
+      [isNil, always(0.5)],
+      [equals<TimelineEvent | undefined>(event), always(0.7)],
+      [T, always(0.2)],
+    ])(annotationHovered);
+
+  const getFill = ({ color, event }: GetColor): string =>
```
Comment (bdauria) [inline www/front_src/src/Resources/Graph/Performance/Graph/useAnnotations.tsx:103]:
```suggestion
  const getFill = ({ color, event }: GetColorProps): string =>
```

**Review by bdauria — DISMISSED** [review]

File: www/front_src/src/Resources/Graph/Performance/Graph/useAnnotations.tsx [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+    return both(gteX, lteX);
+  };
+
+  const changeAnnotationHovered = ({
+    xScale,
+    mouseX,
+    timeline,
+  }: ChangeAnnotationHoveredProps): void => {
+    const isBetweenErorMargin = getIsBetween({
```
Comment (bdauria) [inline www/front_src/src/Resources/Graph/Performance/Graph/useAnnotations.tsx:66]:
```suggestion
    const isWithinErrorMargin = getIsBetween({
```

File: www/front_src/src/Resources/Graph/Performance/Graph/useAnnotations.tsx [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+      xStart: dec(mouseX),
+      xEnd: inc(mouseX),
+    });
+
+    setAnnotationHovered(
+      find(({ startDate, endDate, date }: TimelineEvent) => {
+        if (isNil(startDate)) {
+          return isBetweenErorMargin(xScale(new Date(date)));
```
Comment (bdauria) [inline www/front_src/src/Resources/Graph/Performance/Graph/useAnnotations.tsx:74]:
```suggestion
          return isWithinErrorMargin(xScale(new Date(date)));
```

**Review by bdauria — DISMISSED** [review]

**Review by bdauria — APPROVED** [review]

**Review by kduret — APPROVED** [review]

**Pull request merged**